File: dart/data_structures/graph.dart
// graph.dart

class Graph {
  Map<String, List<String>> _adjacencyList;

  Graph() : _adjacencyList = {};

  void addVertex(String vertex) {
    if (!_adjacencyList.containsKey(vertex)) {
      _adjacencyList[vertex] = [];
    }
  }

  void addEdge(String vertex1, String vertex2) {
    _adjacencyList[vertex1]?.add(vertex2);
    _adjacencyList[vertex2]?.add(vertex1);
  }

  void display() {
    for (var vertex in _adjacencyList.keys) {
      String neighbors = _adjacencyList[vertex]!.join(', ');
      print('$vertex -> $neighbors');
    }
  }
}

void main() {
  Graph myGraph = Graph();

  myGraph.addVertex('A');
  myGraph.addVertex('B');
  myGraph.addVertex('C');
  myGraph.addVertex('D');

  myGraph.addEdge('A', 'B');
  myGraph.addEdge('B', 'C');
  myGraph.addEdge('C', 'D');
  myGraph.addEdge('D', 'A');

  print("Graph Structure:");
  myGraph.display();
}

File: go/data_structures/graph.go
// graph.go

package main

import "fmt"

type Graph struct {
	vertices map[int][]int
}

func NewGraph() *Graph {
	return &Graph{
		vertices: make(map[int][]int),
	}
}

func (g *Graph) AddVertex(vertex int) {
	if _, exists := g.vertices[vertex]; !exists {
		g.vertices[vertex] = []int{}
	}
}

func (g *Graph) AddEdge(vertex1, vertex2 int) {
	g.vertices[vertex1] = append(g.vertices[vertex1], vertex2)
	g.vertices[vertex2] = append(g.vertices[vertex2], vertex1)
}

func (g *Graph) Display() {
	for vertex, neighbors := range g.vertices {
		fmt.Printf("%d -> %v\n", vertex, neighbors)
	}
}

func main() {
	graph := NewGraph()

	graph.AddVertex(1)
	graph.AddVertex(2)
	graph.AddVertex(3)
	graph.AddVertex(4)

	graph.AddEdge(1, 2)
	graph.AddEdge(2, 3)
	graph.AddEdge(3, 4)
	graph.AddEdge(4, 1)

	fmt.Println("Graph:")
	graph.Display()
}
